src/middlewared/middlewared/plugins/smb_/groupmap.py

```diff
@@ -464,3 +469,49 @@ def synchronize_group_mappings(self, job, bypass_sentinel_check=False):
                 self.middleware.call_sync('idmap.gencache.flush')
             except Exception:
                 self.logger.warning('Failed to flush caches after groupmap changes.', exc_info=True)
+
+    @private
+    def migrate_share_groupmap(self):
+        rejects = {g['sid']: g for g in query_groupmap_entries(GroupmapFile.REJECT, [
+            ['entry_type', '=', GroupmapEntryType.GROUP_MAPPING.name]
+        ], {})}
+
+        if not rejects:
+            return
+
+        reject_gids = [g['gid'] for g in rejects.values()]
+
+        set_reject_sids = set(list(rejects.keys()))
+        current = {g['gid']: g['sid'] for g in self.middleware.call_sync('group.query', [
+            ['gid', 'in', reject_gids], ['local', '=', True]
+        ])}
+
+        for sid, gm in rejects.copy().items():
+            rejects[sid]['new_sid'] = current.get(gm['gid'])
+
+        for share in self.middleware.call_sync('sharing.smb.query'):
+            acl = self.middleware.call_sync('smb.sharesec.getacl', share['name'])['share_acl']
+            acl_sids = set([ace['ae_who_sid'] for ace in acl])
+            if not set_reject_sids & acl_sids:
+                continue
+
+            for idx, ace in enumerate(acl):
+                if (entry := rejects.get(ace['ae_who_sid'])) is None:
+                    continue
+
+                if entry['new_sid'] == ace['ae_who_sid']:
+                    # This in theory shouldn't happen, but nothing to change
```

Comment:
Does this indicate a failure in earlier processing?  Maybe log this unexpected condition?